pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,337 @@
+	mapDataSourceRefToPVC := func(obj client.Object) (reqs []reconcile.Request) {
+		var pvcs corev1.PersistentVolumeClaimList
+		matchingFields := client.MatchingFields{dataSourceRefField: indexKeyFunc(obj.GetNamespace(), obj.GetName())}
+		if err := mgr.GetClient().List(context.TODO(), &pvcs, matchingFields); err != nil {
+			log.Error(err, "Unable to list PVCs", "matchingFields", matchingFields)
+			return reqs
+		}
+		for _, pvc := range pvcs.Items {
+			reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}})
+		}
+		return
+	}
+
+	if err := c.Watch(&source.Kind{Type: sourceType},
+		handler.EnqueueRequestsFromMapFunc(mapDataSourceRefToPVC),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcilerBase) getPVCPrime(pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	pvcPrime := &corev1.PersistentVolumeClaim{}
+	pvcPrimeKey := types.NamespacedName{Namespace: pvc.Namespace, Name: PVCPrimeName(pvc)}
+	if err := r.client.Get(context.TODO(), pvcPrimeKey, pvcPrime); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, nil
+	}
+	return pvcPrime, nil
+}
+
+func (r *ReconcilerBase) getPV(volName string) (*corev1.PersistentVolume, error) {
+	pv := &corev1.PersistentVolume{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: volName}, pv); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return pv, nil
+}
+
+func (r *ReconcilerBase) handleStorageClass(pvc *corev1.PersistentVolumeClaim) (bool, bool, error) {
+	if pvc.Spec.StorageClassName == nil {
+		return false, false, nil
+	}
+
+	waitForFirstConsumer := false
+	storageClass, err := cc.GetStorageClassByName(r.client, pvc.Spec.StorageClassName)
+	if err != nil {
+		return false, waitForFirstConsumer, err
+	}
+
+	if checkIntreeStorageClass(pvc, storageClass) {
+		r.log.V(2).Info("can't use populator for PVC with in-tree storage class", "namespace", pvc.Namespace, "name", pvc.Name, "provisioner", storageClass.Provisioner)
+		return false, waitForFirstConsumer, nil
+	}
+
+	if storageClass.VolumeBindingMode != nil && *storageClass.VolumeBindingMode == storagev1.VolumeBindingWaitForFirstConsumer {
+		waitForFirstConsumer = true
+		nodeName := pvc.Annotations[AnnSelectedNode]
+		if nodeName == "" {
+			// Wait for the PVC to get a node name before continuing
+			return false, waitForFirstConsumer, nil
+		}
+	}
+
+	return true, waitForFirstConsumer, nil
+}
+
+func (r *ReconcilerBase) createPVCPrime(pvc *corev1.PersistentVolumeClaim, source client.Object, waitForFirstConsumer bool, pvcModifier pvcModifierFunc) (*corev1.PersistentVolumeClaim, error) {
+	labels := map[string]string{
+		common.CDILabelKey: common.CDILabelValue,
+	}
+	annotations := make(map[string]string)
+	annotations[cc.AnnImmediateBinding] = ""
+	if waitForFirstConsumer {
+		annotations[cc.AnnSelectedNodeName] = pvc.Annotations[AnnSelectedNode]
```

Comment:
Notice how I'm copying the `AnnSelectedNode` from target PVC into `AnnSelectedNodeName`  of PVC'. This annotation is then copied into the importer pod's `NodeName` in import-controller. I noticed your comment in Shelly's PR and tried improving this behavior, but I think I messed something up because it's causing the tests to fail.

Reply:
Once I get that to work I'll add proper comments and maybe change the annotation name so it's not as confusing.